src/org/labkey/test/components/glassLibrary/heatmap/HeatMap.java

```
@@ -0,0 +1,224 @@
+        return Locator.linkWithText(linkText).waitForElement(row, WAIT_FOR_JAVASCRIPT);
+    }
+
+    public WebElement getSummaryLink(String rowLinkText)
+    {
+        WebElement row = getOptionalRow(rowLinkText).get();
+        return Locator.tagWithAttributeContaining("span", "title", " last 12 months")
+                .child("a").findElement(row);
+    }
+
+    public WebElement getCell(String rowLinkText, String headerText)
+    {
+        WebElement row = getOptionalRow(rowLinkText).get();
+        return Locator.css("td[headers=\"" + headerText + "\"]").findElement(row);
+    }
+
+    public String getCellColor(String rowLinkText, String headerText)
+    {
+        WebElement cell = getCell(rowLinkText, headerText);
+        WebElement backgroundCell = Locator.tag("div").findElement(cell);
+        return backgroundCell.getCssValue("background-color");
+    }
+
+    public String getToolTipTextForCell(String rowLinkText, String headerText)
+    {
+        WebElement cell = getCell(rowLinkText, headerText);
+        getWrapper().mouseOver(cell);
+        WebDriverWrapper.waitFor(()-> null != cell.getAttribute("aria-describedby"),
+                "MouseOver did not cause cell to show tooltip", 2000);
+        String toolTipId = cell.getAttribute("aria-describedby");
+        WebElement toolTip = Locator.id(toolTipId).findElement(getDriver());
+        return toolTip.getText();
+    }
+
+    public Boolean isLoaded()
+    {
+        return !_loadingRowLoc.existsIn(this) &&
+                !_spinnerLoc.existsIn(this) &&
+                Locator.tag("td").existsIn(this);
+    }
+
+    public Boolean hasData()
+    {
+        if (!isLoaded())
+            return false;
+        else
+        {
+            try {
+                List<WebElement> elements = _emptyRowLoc.findElements(getComponentElement());
+                return elements.size() == 0;
+            }
+            catch (StaleElementReferenceException e)
```

Do you need to catch NoSuchElement in addition to StaleElement?

---

It probably can't hurt

---

If either of those exceptions is thrown, they will be from the element returned by `getComponentElement`, not the empty row.

---

Yes, I think the try/catch here was vestigial from a time where these methods were handling a row element- so if we hit that exception we probably shouldn't swallow it